frame/support/src/storage/mod.rs

```diff
@@ -811,16 +811,130 @@ mod private {
 
 	impl<T: Encode> Sealed for Vec<T> {}
 	impl<Hash: Encode> Sealed for Digest<Hash> {}
+	impl<T: ValueType, S: Get<usize>> Sealed for BoundedVec<T, S> {}
 }
 
 impl<T: Encode> StorageAppend<T> for Vec<T> {}
 impl<T: Encode> StorageDecodeLength for Vec<T> {}
 
-/// We abuse the fact that SCALE does not put any marker into the encoding, i.e.
-/// we only encode the internal vec and we can append to this vec. We have a test that ensures
-/// that if the `Digest` format ever changes, we need to remove this here.
+/// We abuse the fact that SCALE does not put any marker into the encoding, i.e. we only encode the
+/// internal vec and we can append to this vec. We have a test that ensures that if the `Digest`
+/// format ever changes, we need to remove this here.
 impl<Hash: Encode> StorageAppend<DigestItem<Hash>> for Digest<Hash> {}
 
+use crate::traits::Get;
+use sp_std::convert::TryFrom;
+
+pub trait ValueType: FullCodec + Default {}
+impl<T: FullCodec + Default> ValueType for T {}
+
+#[derive(Encode, Decode, Default)]
+pub struct BoundedVec<T: ValueType, S: Get<usize>>(Vec<T>, sp_std::marker::PhantomData<S>);
+
+impl<T: ValueType, S: Get<usize>> BoundedVec<T, S> {
+	fn unchecked_from(t: Vec<T>) -> Self {
+		Self(t, Default::default())
+	}
+	// TODO: add a subset of the Vec's API here that generally doesn't expand it in any way. Adding
+	// any API that can insert/append/enqueue is tricky and needs care.
```

Comment:
You should always be able to get the length of the vec querying the first bytes of the item.

This same trick here: https://github.com/paritytech/substrate/blob/master/frame/collective/src/lib.rs#L729

Reply:
Also done for `decoded_len`: https://crates.parity.io/frame_support/storage/trait.StorageValue.html#method.decode_len

Reply:
Was thinking about this: probably easier to only provide a limited subset of super useful functions on `BoundedVec`, such as `len`, `get` and `iter`. If you need anything other than this, you call a consuming `fn inner(self) -> Vec<T>` to get the inner `Vec<_>`, do whatever you want with it as if it was a `Vec<_>`, make sure yourself that you don't exceed the bound, and then try and reconstruct the original type with `TryFrom<Vec<T>> for BoundedVec<_, _>`.

Reply:
> Also done for `decode_len`: https://crates.parity.io/frame_support/storage/trait.StorageValue.html#method.decode_len

I already provide `decode_len`, if that's what you are looking for.